File: src/models/state.cpp
#include "state.hpp"

/////////////////////////////////////
//
//	KET VECTORS
//
/////////////////////////////////////

QMDDSuite state::Ket0() {
    return QMDDSuite(QMDDEdge(1.0, make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {edgeZero, edgeZero},
    })));
};

QMDDSuite state::Ket1() {
    return QMDDSuite(QMDDEdge(1.0, make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeZero, edgeZero},
        {edgeOne, edgeZero},
    })));
};

QMDDSuite state::KetPlus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {edgeOne, edgeZero},
    })));
};

QMDDSuite state::KetMinus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {QMDDEdge(-1.0), edgeZero},
    })));
};

QMDDSuite state::KetI() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {QMDDEdge(i), edgeZero},
    })));
};

QMDDSuite state::KetIMinus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {QMDDEdge(-i), edgeZero},
    })));
};


/////////////////////////////////////
//
//	BRA VECTORS
//
/////////////////////////////////////

QMDDSuite state::Bra0() {
    return QMDDSuite(QMDDEdge(1.0, make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeZero},
        {edgeZero, edgeZero}
    })));
};

QMDDSuite state::Bra1() {
    return QMDDSuite(QMDDEdge(1.0, make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeZero, edgeOne},
        {edgeZero, edgeZero}
    })));
};

QMDDSuite state::BraPlus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, edgeOne},
        {edgeZero, edgeZero}
    })));
};

QMDDSuite state::BraMinus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, QMDDEdge(-1.0)},
        {edgeZero, edgeZero}
    })));
};

QMDDSuite state::BraI() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, QMDDEdge(i)},
        {edgeZero, edgeZero}
    })));
};

QMDDSuite state::BraIMinus() {
    return QMDDSuite(QMDDEdge(1.0 / sqrt(2.0), make_shared<QMDDNode>(vector<vector<QMDDEdge>>{
        {edgeOne, QMDDEdge(-i)},
        {edgeZero, edgeZero}
    })));
};

File: src/macOS/Sources/state.swift
import Foundation
import Numerics

enum state {

    /////////////////////////////////////
    //
    //	KET VECTORS
    //
    /////////////////////////////////////

    static func Ket0() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0, node: QMDDNode(edges: [
            [edgeOne],
            [edgeZero]
        ])))
    }

    static func Ket1() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0, node: QMDDNode(edges: [
            [edgeZero],
            [edgeOne]
        ])))
    }

    static func KetPlus() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne],
            [edgeOne]
        ])))
    }

    static func KetMinus() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne],
            [QMDDEdge(weight: -1.0, node: nil)]
        ])))
    }

    static func KetPlusY() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne],
            [QMDDEdge(weight: i, node: nil)]
        ])))
    }

    static func KetMinusY() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne],
            [QMDDEdge(weight: -i, node: nil)]
        ])))
    }

    /////////////////////////////////////
    //
    //	BRA VECTORS
    //
    /////////////////////////////////////

    static func Bra0() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0, node: QMDDNode(edges: [
            [edgeOne, edgeZero]
        ])))
    }

    static func Bra1() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0, node: QMDDNode(edges: [
            [edgeZero, edgeOne]
        ])))
    }

    static func BraPlus() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne, edgeOne]
        ])))
    }

    static func BraMinus() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne, QMDDEdge(weight: -1.0, node: nil)]
        ])))
    }

    static func BraPlusY() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne, QMDDEdge(weight: i, node: nil)]
        ])))
    }

    static func BraMinusY() -> QMDDState {
        return QMDDState(edge: QMDDEdge(weight: 1.0 / sqrt(2.0), node: QMDDNode(edges: [
            [edgeOne, QMDDEdge(weight: -i, node: nil)]
        ])))
    }
}